pymc4/distributions/state_functions.py

```diff
@@ -0,0 +1,250 @@
+    from bernoulli distribution with p=0.5.
+
+    Parameters
+    ----------
+    scale: Union[List[Any], Any]
+        a `Tensor` or Python `list` of `Tensor`s of any shapes and `dtypes`
+        controlling the scale of the proposal distribution.
+    name: Optional[str]
+        Python `str` name prefixed to Ops created by this function.
+        Default value: 'categorical_uniform_fn'.
+    """
+
+    _name = "bernoulli_fn"
+
+    def __init__(self, scale: Union[List[Any], Any] = 1.0, name: Optional[str] = None):
+        super().__init__(name)
+        self.scale = scale
+
+    def _fn(self, state_parts: List[tf.Tensor], seed: Optional[int]) -> List[tf.Tensor]:
+        scale = self.scale
+        with tf.name_scope(self._name or "bernoulli_fn"):
+            scales = scale if mcmc_util.is_list_like(scale) else [scale]
+            if len(scales) == 1:
+                scales *= len(state_parts)
+            if len(state_parts) != len(scales):
+                raise ValueError("`scale` must broadcast with `state_parts`")
+
+            def generate_new_values(state_part, scale_part):
+                # TODO: is there a more elegant way
+                ndim = scale_part.get_shape().ndims
+                reduced_elem = tf.squeeze(
+                    tf.slice(
+                        scale_part, tf.zeros(ndim, dtype=tf.int32), tf.ones(ndim, dtype=tf.int32),
+                    )
+                )
+                delta = tfd.Bernoulli(probs=0.5 * reduced_elem, dtype=state_part.dtype).sample(
+                    seed=seed, sample_shape=(tf.shape(state_part))
+                )
+                state_part += delta
+                state_part = state_part % 2.0
+                return state_part
+
+            state_parts_st = tf.stack(state_parts)
+            orig_dtype = state_parts_st.dtype
+            # TODO: we create scale_part with shape=state_part.shape
+            # each function call. But scalar value would be enough
+            # The issue is that we pass the single tensor to map_fn
+            # and we need the semantics of the tensors to be the same
+            # to be able to stack them together.
+
+            shape_ = state_parts_st.shape
+            inds_tile = tf.concat(
+                [tf.constant([shape_[0]]), tf.ones(shape_.ndims, dtype=tf.int32)], axis=0,
+            )
+            shape_ = (1,) + shape_[1:] + (shape_[0],)
+            scales = tf.tile(tf.broadcast_to(scales, shape_), inds_tile)[..., 0]
+            state_parts_st = tf.cast(state_parts_st, dtype=scales.dtype)
+            deltas = tf.unstack(
+                tf.map_fn(
+                    lambda x: generate_new_values(
+                        x[0], x[1]
+                    ),  # TODO: some issues with unpack and tf graph
+                    tf.stack([state_parts_st, scales], axis=1),
+                )
+            )
+            return tf.unstack(tf.cast(deltas, dtype=orig_dtype))
```

Comment:
Could you walk me through why scale is needed here? Because otherwise it is just a Bernouli random with probs=.5 and shape=state_part